azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+                            initial_register_op.complete(error=error)
+                        elif registration_status == "assigning":
+
+                            def copy_result_to_original_op(op, error):
+                                logger.debug(
+                                    "Copying registration result from Query Status Op to Registration Op"
+                                )
+                                initial_register_op.registration_result = op.registration_result
+                                initial_register_op.error = error
+
+                            query_worker_op = initial_register_op.spawn_worker_op(
+                                worker_op_type=pipeline_ops_provisioning.SendQueryRequestOperation,
+                                request_payload=" ",
+                                operation_id=operation_id,
+                                callback=copy_result_to_original_op,
+                            )
+
+                            self.send_op_down(query_worker_op)
+                        else:
+                            error = exceptions.ServiceError(
+                                "Registration Request encountered an invalid registration status {status} with a status code of {status_code}".format(
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+                            initial_register_op.complete(error=error)
+
+                else:
+                    initial_register_op.complete(error=error)
+
+            registration_payload = DeviceRegistrationPayload(
+                registration_id=initial_register_op.registration_id,
+                custom_payload=initial_register_op.request_payload,
+            )
+            self.send_op_down(
+                pipeline_ops_base.RequestAndResponseOperation(
+                    request_type=constant.REGISTER,
+                    method="PUT",
+                    resource_location="/",
+                    request_body=registration_payload.get_json_string(),
+                    callback=on_registration_response,
+                )
+            )
+
+        else:
+            super(RegistrationStage, self)._run_op(operation)
+
+
+class ProvisioningTimeoutStage(PipelineStage):
+    """
+    The purpose of the timeout stage is to add timeout errors to select operations
+
+    The timeout_intervals attribute contains a list of operations to track along with
+    their timeout values.
+
+    For each operation that needs a timeout check, this stage will add a timer to
+    the operation.  If the timer elapses, this stage will fail the operation with
+    a PipelineTimeoutError.
+    """
+
+    def __init__(self):
+        super(ProvisioningTimeoutStage, self).__init__()
+        self.timeout_intervals = {
+            pipeline_ops_base.RequestOperation: constant.DEFAULT_TIMEOUT_INTERVAL
```

Comment:
>RequestOperation [](start = 30, length = 16)

I suspect that this stage isn't doing what you think it's doing.  The way I read this code, it is causing a timeout error if the request fails to publish.  However, if the request publishes successfully, but no response is received, then this stage does nothing.

Timing out on higher level operations like this is tricky.  I know that the node code had the concept of a registration operation timing out, but that code is weak and I don't think it has any real benefits. (I'm bad-mouthing my own code here since I originally wrote that.)

Reply:
I don't think we have any high level timeout in iothub, even for request and response operations like `get_twin`. Do we need this here? And if so, do we also need it in iothub? If it's also needed in iothub, is there some kind of more generic shared solution we could implement?

---
In reply to: [364419178](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r364419178) [](ancestors = 364419178)

Reply:
I think we do need it in iothub, but it's tricky because it starts to quickly descend into the topic of cancelling operations, and that's hard to get right, especially since the underlying protocol doesn't really help us here.  

---
In reply to: [364529899](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r364529899) [](ancestors = 364529899,364419178)

Reply:
i have debugged this. The timing out part is fine but the clearing of the timer happens when the request operation completes which is wrong.
The way i see it , clearing of timer should only happen once we receive response from service. 

the only response handlers/callbacks that I have access to and which are specific to DPS are the `on_registration_response` and `on_query_response` functions as other responses are handled by generic stages in `pipeline_stages_base.py`. 

So I have omitted the `ProvisioningTimeoutStage` completely and have included timeout timers in the `RegistrationStage` and `PollingStatusStage` now.

if we want to included time out for all responses from service may be the `CoordinateRequestAndResponseStage` is a better stage to put the timer.

---
In reply to: [365998937](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365998937) [](ancestors = 365998937,364529899,364419178)